### PR TITLE
Adding GPU acceleration to encode_jpeg

### PR DESCRIPTION
**Pull request by deekay42**
Summary:
I'm adding GPU support to the existing torchvision.io.encode_jpeg function. If the input tensors are on the GPU, the CUDA version will be used and the CPU version otherwise. 

Performance numbers indicate over 5000 imgs/s on 1 A100 GPU:
```
Processor: x86_64
Platform: Linux-5.12.0-0_fbk7_zion_6511_gd766966f605a-x86_64-with-glibc2.34
Logical CPUs: 192

CUDA device: NVIDIA PG509-210
Total Memory: 84.99 GB

Mean image size: 551x676
[----------------------------------------------------- Image Encoding -----------------------------------------------------]
                                                                                  |  1 images  |  100 images  |  1000 images
1 threads: -----------------------------------------------------------------------------------------------------------------
      CPU (unfused): [torchvision.io.encode_jpeg(img) for img in batch_input]     |   2466.1   |   219121.6   |   2169960.9
      CPU (fused): torchvision.io.encode_jpeg(batch_input)                        |   2627.3   |   221350.3   |   2098801.9
      CUDA:7 (unfused): [torchvision.io.encode_jpeg(img) for img in batch_input]  |    256.8   |    21060.6   |    212853.1
      CUDA:7 (fused): torchvision.io.encode_jpeg(batch_input)                     |    223.8   |    16829.9   |    193673.9
12 threads: ----------------------------------------------------------------------------------------------------------------
      CPU (unfused): [torchvision.io.encode_jpeg(img) for img in batch_input]     |   2512.9   |   216763.1   |   2161373.3
      CPU (fused): torchvision.io.encode_jpeg(batch_input)                        |   2608.3   |   223391.8   |   2152523.7
      CUDA:7 (unfused): [torchvision.io.encode_jpeg(img) for img in batch_input]  |    220.8   |    24009.1   |    245279.8
      CUDA:7 (fused): torchvision.io.encode_jpeg(batch_input)                     |    219.3   |    16971.8   |    175770.5
24 threads: ----------------------------------------------------------------------------------------------------------------
      CPU (unfused): [torchvision.io.encode_jpeg(img) for img in batch_input]     |   2549.4   |   215578.0   |   2195501.8
      CPU (fused): torchvision.io.encode_jpeg(batch_input)                        |   2427.9   |   225574.3   |   2139114.4
      CUDA:7 (unfused): [torchvision.io.encode_jpeg(img) for img in batch_input]  |    219.9   |    21726.3   |    214842.8
      CUDA:7 (fused): torchvision.io.encode_jpeg(batch_input)                     |    212.6   |    17351.6   |    172555.7

Times are in microseconds (us).```

Test Plan:
1. pytest test -vvv
2. ufmt format torchvision
3. flake8 torchvision

Reviewers:

Subscribers:

Tasks:

Tags:

<!-- Before submitting a PR, please make sure to check our contributing guidelines regarding code formatting, tests, and documentation: https://github.com/pytorch/vision/blob/main/CONTRIBUTING.md -->

### CODE REVIEW

File: test/test_image.py [hunk 1]
```diff
@@ -67,6 +67,155 @@ def normalize_dimensions(img_pil):
...
+        # For more detail as to why check out: https://github.com/NVIDIA/cuda-samples/issues/23#issuecomment-559283013
+    if not contiguous:
+        decoded_image_tv = decoded_image_tv.permute(1, 2, 0).contiguous().permute(2, 0, 1)
+    encoded_jpeg_cuda_tv = encode_fn(decoded_image_tv.cuda(), quality=75)
+    decoded_jpeg_cuda_tv = decode_jpeg(encoded_jpeg_cuda_tv.cpu())
+
+    # the actual encoded bytestreams from libnvjpeg and libjpeg-turbo differ for the same quality
+    # instead, we re-decode the encoded image and compare to the original
+    abs_mean_diff = (
+        (decoded_jpeg_cuda_tv.type(torch.float32) - decoded_image_tv.type(torch.float32)).abs().mean().item()
```
Comment (NicolasHug) [inline test/test_image.py:94]: For consistency with the rest of the code and to be more succinct, we more commonly convert to float using `.float()`:
```suggestion
        (decoded_jpeg_cuda_tv.float() - decoded_image_tv.float()).abs().mean().item()
```

(or by calling `.to(torch.float))` instead of `type()`)

File: test/test_image.py [hunk 2]
```diff
@@ -67,6 +67,155 @@ def normalize_dimensions(img_pil):
...
+
+    encode_fn = torch.jit.script(encode_jpeg) if scripted else encode_jpeg
+
+    decoded_images_tv_cuda = [img.cuda() for img in decoded_images_tv]
+    encoded_jpegs_tv_cuda = encode_fn(decoded_images_tv_cuda, quality=75)
+    decoded_jpegs_tv_cuda = [decode_jpeg(img.cpu()) for img in encoded_jpegs_tv_cuda]
+
+    for original, encoded_decoded in zip(decoded_images_tv, decoded_jpegs_tv_cuda):
+        c, h, w = original.shape
+        abs_mean_diff = (original.type(torch.float32) - encoded_decoded.type(torch.float32)).abs().mean().item()
```
Comment (NicolasHug) [inline test/test_image.py:122]: same as above

File: torchvision/transforms/v2/functional/_augment.py [hunk 3]
```diff
@@ -78,9 +78,14 @@ def jpeg_image(image: torch.Tensor, quality: int) -> torch.Tensor:
...
-    image = torch.stack(image, dim=0).view(original_shape)
-    return image
+    images = []
+    for i in range(image.shape[0]):
+        encoded_image = encode_jpeg(image[i], quality=quality)
+        assert isinstance(encoded_image, torch.Tensor)
+        images.append(decode_jpeg(encoded_image))
+
+    images = torch.stack(images, dim=0).view(original_shape)
+    return images
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_augment.py:88]: Were these changes necessary or was this just for debugging?

File: torchvision/io/image.py [hunk 4]
```diff
@@ -218,11 +232,14 @@ def write_jpeg(input: torch.Tensor, filename: str, quality: int = 75):
     if not torch.jit.is_scripting() and not torch.jit.is_tracing():
         _log_api_usage_once(write_jpeg)
     output = encode_jpeg(input, quality)
+    assert isinstance(output, torch.Tensor)
```
Comment (NicolasHug) [inline torchvision/io/image.py:235]: This can probably be removed?

File: torchvision/io/image.py [hunk 5]
```diff
@@ -180,27 +183,38 @@ def decode_jpeg(
...
 
-def encode_jpeg(input: torch.Tensor, quality: int = 75) -> torch.Tensor:
+def encode_jpeg(
+    input: Union[torch.Tensor, List[torch.Tensor]], quality: int = 75
+) -> Union[torch.Tensor, List[torch.Tensor]]:
     """
-    Takes an input tensor in CHW layout and returns a buffer with the contents
-    of its corresponding JPEG file.
+    Takes a (list of) input tensor(s) in CHW layout and returns a (list of) buffer(s) with the contents
+    of the corresponding JPEG file(s).
```
Comment (NicolasHug) [inline torchvision/io/image.py:191]:
```suggestion
    of the corresponding JPEG file(s).
    
    .. note::

        Passing a list of CUDA tensors is more efficient than repeated individual calls to ``encode_jpeg``. 
        For CPU tensors the performance is equivalent.
```

File: torchvision/io/image.py [hunk 6]
```diff
@@ -180,27 +183,38 @@ def decode_jpeg(
...
-        output (Tensor[1]): A one dimensional int8 tensor that contains the raw bytes of the
-            JPEG file.
+        output (Tensor[1] or list[Tensor[1]]): A (list of) one dimensional uint8 tensor(s) that contain the raw bytes of the JPEG file.
     """
     if not torch.jit.is_scripting() and not torch.jit.is_tracing():
         _log_api_usage_once(encode_jpeg)
     if quality < 1 or quality > 100:
         raise ValueError("Image quality should be a positive number between 1 and 100")
+    if isinstance(input, list):
+        assert len(input) > 0, "encode_jpeg requires at least one input tensor when a list is passed"
```
Comment (NicolasHug) [inline torchvision/io/image.py:207]: We typcially raise ValueError for such user-input issues:

```suggestion
        if not input:
            raise ValueError("encode_jpeg requires at least one input tensor when a list is passed")

File: torchvision/io/image.py [hunk 7]
```diff
@@ -180,27 +183,38 @@ def decode_jpeg(
...
+        if input[0].device.type == "cuda":
+            return torch.ops.image.encode_jpeg_cuda(input, quality)
+        else:
+            return [torch.ops.image.encode_jpeg(image, quality) for image in input]
```
Comment (NicolasHug) [inline torchvision/io/image.py:211]: Looks like passing a list of CPU tensors is untested (otherwise it would have failed on this line). Let's add small smoke test for that (we can just randomly generate an image instead of reading from the existing jpegs). We just need to assert that this runs without errors.

```suggestion
            return [torch.ops.image.encode_jpeg(input, quality) for image in input]
```

File: torchvision/csrc/io/image/cuda/encode_jpeg_cuda.cpp [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        image.dim() == 3 && image.numel() > 0,
+        "Input data should be a 3-dimensional tensor");
+
+    TORCH_CHECK(
+        image.size(0) == 3,
+        "The number of channels should be 3, got: ",
+        image.size(0));
+
+    // nvjpeg requires images to be contiguous
+    contig_images.push_back(image.contiguous());
```
Comment (NicolasHug) [inline torchvision/csrc/io/image/cuda/encode_jpeg_cuda.cpp:92]: Q: Assuming `image` was already contiguous, does this lead to an unnecessary copy?

File: torchvision/csrc/io/image/cuda/encode_jpeg_cuda.cpp [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+  int channels = src_image.size(0);
+  int height = src_image.size(1);
+  int width = src_image.size(2);
+
+  nvjpegStatus_t samplingSetResult = nvjpegEncoderParamsSetSamplingFactors(
+      nv_enc_params, NVJPEG_CSS_444, stream);
+  TORCH_CHECK(
+      samplingSetResult == NVJPEG_STATUS_SUCCESS,
+      "Failed to set nvjpeg encoder params sampling factors: ",
+      samplingSetResult);
```
Comment (NicolasHug) [inline torchvision/csrc/io/image/cuda/encode_jpeg_cuda.cpp:159]: In https://github.com/pytorch/vision/pull/8375#discussion_r1571016848 you mentioned this may not be needed anymore?

**Review by NicolasHug — COMMENTED** [review]
Thanks a lot @deekay42 . I made another pass but this looks good!

File: test/test_image.py [hunk 10]
```diff
@@ -67,6 +67,155 @@ def normalize_dimensions(img_pil):
...
 
 
+@needs_cuda
+@pytest.mark.parametrize(
+    "img_path",
+    [pytest.param(jpeg_path, id=_get_safe_image_name(jpeg_path)) for jpeg_path in get_images(IMAGE_ROOT, ".jpg")],
+)
+@pytest.mark.parametrize("scripted", (False, True))
+@pytest.mark.parametrize("contiguous", (False, True))
+def test_single_encode_jpeg_cuda(img_path, scripted, contiguous):
```
[the remark this answers is not in the document]
Comment (deekay42) [reply test/test_image.py:77]: done

File: test/test_image.py [hunk 11]
```diff
@@ -67,6 +67,155 @@ def normalize_dimensions(img_pil):
...
+    decoded_image_tv = read_image(img_path)
+    encode_fn = torch.jit.script(encode_jpeg) if scripted else encode_jpeg
+
+    if "cmyk" in img_path:
+        pytest.xfail("Encoding a CMYK jpeg isn't supported")
+    if decoded_image_tv.shape[0] == 1:
+        pytest.xfail("Decoding a grayscale jpeg isn't supported")
+        # For more detail as to why check out: https://github.com/NVIDIA/cuda-samples/issues/23#issuecomment-559283013
+    if not contiguous:
+        decoded_image_tv = decoded_image_tv.permute(1, 2, 0).contiguous().permute(2, 0, 1)
```
[the remark this answers is not in the document]
Comment (deekay42) [reply test/test_image.py:87]: done

File: test/test_image.py [hunk 12]
```diff
@@ -67,6 +67,155 @@ def normalize_dimensions(img_pil):
...
+def test_batch_encode_jpegs_cuda(scripted, contiguous):
+    decoded_images_tv = []
+    for jpeg_path in get_images(IMAGE_ROOT, ".jpg"):
+        if "cmyk" in jpeg_path:
+            continue
+        decoded_image = read_image(jpeg_path)
+        if decoded_image.shape[0] == 1:
+            continue
+        if not contiguous:
+            decoded_image = decoded_image.permute(1, 2, 0).contiguous().permute(2, 0, 1)
```
[the remark this answers is not in the document]
Comment (deekay42) [reply test/test_image.py:111]: done

File: test/test_image.py (hunk 1 above)
Comment (deekay42) [reply test/test_image.py:94]: done

File: torchvision/transforms/v2/functional/_augment.py (hunk 3 above)
Comment (deekay42) [reply torchvision/transforms/v2/functional/_augment.py:88]: These changes were necessary. The return type of encode_jpeg changed from torch.Tensor to Union[list[torch.Tensor], torch.Tensor]. Without this assertion, the mypy will complain.

File: torchvision/io/image.py (hunk 4 above)
Comment (deekay42) [reply torchvision/io/image.py:235]: No, same as above.

File: torchvision/io/image.py (hunk 5 above)
Comment (deekay42) [reply torchvision/io/image.py:191]: done

File: torchvision/io/image.py (hunk 6 above)
Comment (deekay42) [reply torchvision/io/image.py:207]: done

File: torchvision/io/image.py (hunk 7 above)
Comment (deekay42) [reply torchvision/io/image.py:211]: Agreed on adding a little smoke test for CPU batch encode, but the code here doesn't look wrong to me.

File: torchvision/csrc/io/image/cuda/encode_jpeg_cuda.cpp (hunk 8 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/encode_jpeg_cuda.cpp:92]: GPT4 and Llama3 think that .contiguous() just returns the original tensor if it's already contiguous, but I couldn't find any official confirmation in the libtorch docs. At any rate, I included some explicit validation.

File: torchvision/csrc/io/image/cuda/encode_jpeg_cuda.cpp (hunk 9 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/encode_jpeg_cuda.cpp:159]: It's again one of those things that's not mentioned in the docs, but apparently is mandatory. See here: https://stackoverflow.com/questions/65929613/nvjpeg-encode-packed-bgr

File: torchvision/io/image.py (hunk 7 above)
Comment (NicolasHug) [reply torchvision/io/image.py:216]: yes sorry I was reading this wrong

File: torchvision/csrc/io/image/cuda/encode_decode_jpeg_cuda.h [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace vision {
+namespace image {
+
+C10_EXPORT torch::Tensor decode_jpeg_cuda(
+    const torch::Tensor& data,
+    ImageReadMode mode,
+    torch::Device device);
+
+C10_EXPORT std::vector<torch::Tensor> encode_jpeg_cuda(
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/encode_decode_jpeg_cuda.h:21]: Out of curiosity, why does this interface take in and return a std::vector<Tensor> instead of a single stacked  Tensor? I am asking because in the TorchMedia decoder, our users wanted a single stacked tensor for all the video frames and doing the torch.stack() operation on a python list of tensors was time-consuming.

[the remark this answers is not in the document]
Comment (NicolasHug) [reply torchvision/csrc/io/image/cuda/encode_decode_jpeg_cuda.h:21]: Just chiming in, these are good questions

We need to return a vector instead of a stacked tensor because the encoded jpegs aren't necessarily of the same length. We wouldn't be able to fit those in a stacked tensor since by definition all the tensors in a stack must have the same shape.

As for why we're using a vector for the *input*: this is for similar reason, i.e. the input images may not all be of the same shape. I do think we should allow the Python API (`image.io.decode_jpeg()`) to support a stacked tensor of shape `NCHW`, and possibly convert that internally to a vec of vectors of shape `CHW` so it can be passed to this C++ function, but that can be left out for another PR and I didn't want to bother Dominik with that right now.

[the remark this answers is not in the document]
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/encode_decode_jpeg_cuda.h:21]: Nicolas beat me to it, but yes, the reason for the vector is due to heterogeneous input and output tensor shapes.

File: torchvision/csrc/io/image/cuda/encode_decode_jpeg_cuda.h [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    const std::vector<torch::Tensor>& images,
+    const int64_t quality);
+
+void nvjpeg_init();
```
Comment (NicolasHug) [inline torchvision/csrc/io/image/cuda/encode_decode_jpeg_cuda.h:25]: Since we're not exposing this one, should we put it in a different namespace than in `vision::image`?

File: torchvision/csrc/io/image/cuda/encode_jpeg_cuda.cpp [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+std::vector<torch::Tensor> encode_jpeg_cuda(
+    const std::vector<torch::Tensor>& images,
+    const int64_t quality) {
+  TORCH_CHECK(
+      false, "decode_jpeg_cuda: torchvision not compiled with nvJPEG support");
+}
+
+#else
+
+void nvjpeg_init() {
```
Comment (NicolasHug) [inline torchvision/csrc/io/image/cuda/encode_jpeg_cuda.cpp:35]: This probably doesn't matter too much but `nvjpeg_init()` is declared in `encode_decode_jpeg_cuda.h` no matter what `NVJPEG_FOUND` is, but it is only defined here if `NVJPEG_FOUND` is defined.

File: torchvision/csrc/io/image/cuda/encode_jpegs_cuda.cpp [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+      delete cudaJpegEncoder.release();
+
+    cudaJpegEncoder = std::make_unique<CUDAJpegEncoder>(device);
+
+    // Unfortunately, we cannot rely on the smart pointer releasing the encoder
+    // object correctly upon program exit. This is because, when cudaJpegEncoder
+    // gets destroyed, the CUDA runtime may already be shut down, rendering all
+    // destroy* calls in the encoder destructor invalid. Instead, we use an
+    // atexit hook which executes after main() finishes, but before CUDA shuts
+    // down when the program exits.
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/encode_jpegs_cuda.cpp:62]: What's the guarantee that CUDA shuts down before us?

AFAICt, std::atexit runs these functions in reverse order of when they are called. Is CUDA using atexit() also? If so we need to make sure that is registered before us.

If so, add a comment to that effect.

File: torchvision/csrc/io/image/cuda/encode_jpegs_cuda.h [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+namespace image {
+
+class CUDAJpegEncoder {
+ public:
+  CUDAJpegEncoder(const torch::Device& device);
+  ~CUDAJpegEncoder();
+
+  torch::Tensor encode_jpeg(const torch::Tensor& src_image);
+
+  void setQuality(const int64_t);
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/encode_jpegs_cuda.h:20]: Add a parameter name similar to encode_jpeg above?

File: torchvision/csrc/io/image/cuda/encode_jpegs_cuda.h [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace vision {
+namespace image {
+
+class CUDAJpegEncoder {
+ public:
+  CUDAJpegEncoder(const torch::Device& device);
+  ~CUDAJpegEncoder();
+
+  torch::Tensor encode_jpeg(const torch::Tensor& src_image);
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/encode_jpegs_cuda.h:18]: The name here has underscores while below is using camelCase. Make them consistent?

File: torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h [hunk 19]
```diff
@@ -11,5 +12,9 @@ C10_EXPORT torch::Tensor decode_jpeg_cuda(
     ImageReadMode mode,
     torch::Device device);
 
+C10_EXPORT std::vector<torch::Tensor> encode_jpegs_cuda(
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h:15]: Add a comment here or somewhere for the user to say that it only supports contiguous tensors?

File: torchvision/csrc/io/image/cuda/encode_jpegs_cuda.cpp [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+    auto encoded_image = cudaJpegEncoder->encode_jpeg(image);
+    encoded_images.push_back(encoded_image);
+  }
+
+  // We use a dedicated stream to do the encoding and even though the results
+  // may be ready on that stream we cannot assume that they are also available
+  // on the current stream of the calling context when this function returns. We
+  // use a blocking event to ensure that this is indeed the case. Crucially, we
+  // do not want to block the host (which is what cudaStreamSynchronize would
+  // do) Events allow us to synchronize the streams without blocking the host
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/encode_jpegs_cuda.cpp:107]: Add periods here for punctuation.

Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/encode_jpegs_cuda.cpp:107]: I don't understand this comment.

You are saying we are not blocking the host -- yet I do see there is a cudaEventSynchronize() call in encode_jpeg(). So it appears you are pausing the host every iteration of the for loop. Why does the comment say we are not blocking the host?

**Review by ahmadsharif1 — COMMENTED** [review]
lgtm module nits. And sorry I don't understand the comment about not waiting for each image when the code seems to wait for every image.

File: torchvision/csrc/io/image/cuda/encode_jpegs_cuda.cpp (hunk 16 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/encode_jpegs_cuda.cpp:62]: There is no guarantee. There are a few mentions of using atexit in nvidia forums and stackoverflow https://forums.developer.nvidia.com/t/correct-placement-of-cudadevicereset-for-large-c-application/41104 
https://stackoverflow.com/questions/19184865/cuda-context-destruction-at-host-process-termination
but CUDA shutdown in general is kept quite vague. Everything works fine when on my machine which means CUDA is indeed shutting down after atexit handlers are being called, but I'm adding some additional logic for good measure to make sure that if CUDA is already shut down we don't attempt to run cleanup.

File: torchvision/csrc/io/image/cuda/encode_jpegs_cuda.h (hunk 18 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/encode_jpegs_cuda.h:18]: yup

File: torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h (hunk 19 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h:15]: Line 87 in encode_jpegs_cuda.cpp should takes care of handling non-contiguous images.

File: torchvision/csrc/io/image/cuda/encode_jpegs_cuda.cpp (hunk 20 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/encode_jpegs_cuda.cpp:107]: .

Comment (deekay42) [reply torchvision/csrc/io/image/cuda/encode_jpegs_cuda.cpp:107]: It's a micro-optimization. At certain points during the execution of the overall operator we have to synchronize because there is simply no other way, but at this particular point we only need to sync with the current stream and not the host itself.

**Comment by deekay42** [comment]
> lgtm module nits. And sorry I don't understand the comment about not waiting for each image when the code seems to wait for every image.

Thanks for the review!
Sorry, which comment are you referring to, specifically?

**Review by NicolasHug — APPROVED** [review]
Thanks a ton for the great work @deekay42 !
